lib/auth/auth.go

```diff
@@ -1170,6 +1170,88 @@ func (a *Server) GetMFADevices(ctx context.Context, req *proto.GetMFADevicesRequ
 	}, nil
 }
 
+// verifyMFARespAndAddDeviceWithToken validates MFA register response and on success,
+// adds the new MFA device. This method is specifically for secrets and storage keys associated with a user token.
+func (a *Server) verifyMFARespAndAddDeviceWithToken(ctx context.Context, regResp *proto.MFARegisterResponse, token types.UserToken, deviceName string) error {
+	cap, err := a.GetAuthPreference(ctx)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	// Validate MFARegisterResponse and upsert the new device on success.
+	var dev *types.MFADevice
+	switch resp := regResp.GetResponse().(type) {
+	case *proto.MFARegisterResponse_TOTP:
+		if cap.GetSecondFactor() == constants.SecondFactorU2F {
+			return trace.BadParameter("cluster only allows U2F for second factor")
```

Comment:
This is already broken in face of constants.SecondFactorWebauthn.

Maybe:

```suggestion
		if sf := cap.GetSecondFactor(); sf != constants.SecondFactorOTP && sf != constants.SecondFactorOn && sf != constants.SecondFactorOptional {
			return trace.BadParameter("second factor OTP not allowed by cluster")
```

I would also question whether this is necessary at all - they won't be able to register a TOTP for which we don't already have a secret stored.

Reply:
as discussed here: https://github.com/gravitational/teleport/pull/8103#discussion_r702122249, it would be necessary to prevent users from locking themselves out in case of `reset/invite`